src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,350 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if !NETSTANDARD2_0
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.Loader;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Internal.Log;
+
+namespace Microsoft.CodeAnalysis.CustomMessageHandler;
+
+[ExportWorkspaceService(typeof(ICustomMessageHandlerService)), Shared]
+internal sealed class CustomMessageHandlerService : ICustomMessageHandlerService, IDisposable
+{
+    /// <summary>
+    /// Extensions assembly load contexts and loaded handlers, indexed by handler file path. The handlers are indexed by type name.
+    /// </summary>
+    private readonly Dictionary<string, CustomMessageHandlerExtension> _extensions = new();
```

Comment:
mention concurrency requirements (like if it needs to be locked).

Reply:
will have to undrestand better what a CustomMessageHandlerExtension is.  Note: we have helpers like IsolatedAnalyzerReferenceSet and ANalyzerFileReference which  we might be able to repurpose to help here.

Reply:
Added the following comment to `_lockObject`:
```
// Used to protect access to _extensions, _handlers and _documentHandlers.

Reply:
wfm.

Reply:
It's there now. Maybe I hadn't pushed my changes yet